packages/legend-query-builder/src/components/QueryBuilder.tsx

```diff
@@ -367,10 +371,70 @@ export const QueryBuilder = observer(
       }
       return null;
     };
+
+    const undo = useCallback((): void => {
+      queryBuilderState.changeHistoryState.undo();
+    }, [queryBuilderState.changeHistoryState]);
+
+    const redo = useCallback((): void => {
+      queryBuilderState.changeHistoryState.redo();
+    }, [queryBuilderState.changeHistoryState]);
+
+    useEffect(() => {
+      // this condition is for passing all exisitng tests because when we initialize a queryBuilderState for a test,
+      // we use an empty RawLambda with an empty class and this useEffect is called earlier than initializeWithQuery()
+      if (
+        (queryBuilderState.isQuerySupported && queryBuilderState.class) ||
+        !queryBuilderState.isQuerySupported
+      ) {
+        queryBuilderState.changeHistoryState.cacheNewQuery(
+          queryBuilderState.buildQuery(),
+        );
+      }
+    }, [queryBuilderState, queryBuilderState.hashCode]);
+
+    // bind ctl + z to undo, ctl + y to redo
+    useEffect(() => {
+      const onCtrlZ = (event: KeyboardEvent): void => {
+        if (
```

Comment:
Have you tried registering the commands in queryBuilderState instead of binding them like this?